### PR TITLE
Sindhu/bfloat16 op tests

### PR DESCRIPTION
**Pull request by sindhu-nervana**
This PR implements new tests listed below to test bfloat16 conversion
1. test_conv2d.py 
2. test_conv2dbackpropfilter_nchw.py
3. test_conv2dbackpropfilter_nhwc.py
4. test_conv2dbackpropinput_nchw.py
5. test_conv2dbackpropinput_nhwc.py

### CODE REVIEW

File: test/python/bfloat16/test_conv2d.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+W = 5
+C = 1
+
+filter = np.random.rand(1, 1, 1, 2)
+input_size_nhwc = [N, H, W, C]
+input_size_nchw = [N, C, H, W]
+input_nhwc = tf.placeholder(tf.float32, shape=input_size_nhwc, name='x')
+input_nchw = tf.placeholder(tf.float32, shape=input_size_nchw, name='x')
+
+n_np = np.random.rand(1, 1, 3, 5).astype('f')
```
Comment (shresthamalik) [inline test/python/bfloat16/test_conv2d.py]: use N,H,W,C

File: test/python/bfloat16/test_conv2d.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+# Filter shape is allowed
+# If any fail, then we should place Op on CPU for compute
+
+#Inputs
+N = 1
+H = 3
+W = 5
+C = 1
+
+filter = np.random.rand(1, 1, 1, 2)
```
Comment (shresthamalik) [inline test/python/bfloat16/test_conv2d.py]: Please check if filter needs to be casted to bfloat

File: test/python/bfloat16/test_conv2dbackpropfilter_nchw.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+
+
+#Ngraph Graph
+def ng_model(padding):
+    t1 = tf.placeholder(dtype=tf.float32, shape=input_sizes_nchw, name='t1')
+    t2 = tf.constant(filter_size_hwio, dtype=tf.int32, name='t1')
+    t3 = tf.placeholder(
+        dtype=tf.float32, shape=out_backprop_in_sizes[padding], name='t3')
+
+    filt = nn_ops.conv2d_backprop_filter(
```
Comment (shresthamalik) [inline test/python/bfloat16/test_conv2dbackpropfilter_nchw.py]: do we need to transpose the output of either tf or ngraph before comparing?

File: test/python/bfloat16/test_conv2dbackpropinput_nhwc.py [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+np.random.seed(5)
+#Inputs
+N = 1
+H = 7
+W = 6
+C = 2
+
+I = C
+O = 2
+FW = 3
```
Comment (shresthamalik) [inline test/python/bfloat16/test_conv2dbackpropinput_nhwc.py]: Please change the variable names like filter/input/FW etc

**Comment by shresthamalik** [comment]
Tests are quite neat, Thanks! 👍

File: test/python/bfloat16/test_conv2dbackpropfilter_nchw.py (hunk 3 above)
Comment (sindhu-nervana) [reply test/python/bfloat16/test_conv2dbackpropfilter_nchw.py]: No, output is same shape as filter which is a constant. the shapes are same for both TF and Ngraph graph

File: test/python/bfloat16/test_conv2d.py (hunk 2 above)
Comment (sindhu-nervana) [reply test/python/bfloat16/test_conv2d.py]: thanks for pointing this out. yes, it does. changed it

File: test/python/bfloat16/test_conv2d.py [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+import os
+
+import ngraph_bridge
+
+#Test Ngraph Op Convolution, TF Op:conv2d
+# Implemented based on NNP's unit test TEST(test_assign_layout, convolution_special_case)
+
+np.random.seed(5)
+
+# Colvolution Op is placed on NNP and conerted to bfloat16 only for the special case below, otherwise it falls back to CPU for compute
```
Comment (kanvi-nervana) [inline test/python/bfloat16/test_conv2d.py]:
```suggestion
# Colvolution Op is placed on NNP and conerted to bfloat16 only for the special case below, 
# otherwise it falls back to CPU for compute
```

Comment (sindhu-nervana) [inline test/python/bfloat16/test_conv2d.py]:
```suggestion
# Colvolution Op is placed on NNP and conerted to 
# bfloat16 only for the special case below, otherwise it falls 
# back to CPU for compute
```

File: test/python/bfloat16/test_conv2d.py [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+t_np = np.transpose(n_np, (0, 2, 3, 1))
+
+
+#TF graph
+def tf_model():
+    stride_nhwc = [1, 2, 2, 1]
+    x = tf.cast(input_nhwc, dtype=tf.bfloat16)
+    filter_cast = tf.cast(filter_size, dtype=tf.bfloat16)
+    m = tf.nn.conv2d(
+        x, filter_cast, stride_nhwc, "SAME", data_format="NHWC", name="m")
```
Comment (kanvi-nervana) [inline test/python/bfloat16/test_conv2d.py]: Should we also have a test where `padding = VALID` ?

File: test/python/bfloat16/test_conv2d.py [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+
+#Ngraph graph
+def ng_model():
+    stride_nchw = [1, 1, 2, 2]
+    m = tf.nn.conv2d(
+        input_nchw,
+        filter_size,
+        stride_nchw,
+        "SAME",
+        data_format="NCHW",
```
Comment (kanvi-nervana) [inline test/python/bfloat16/test_conv2d.py]: Do we need a test for `data_format=NHWC`?

File: test/python/bfloat16/test_conv2dbackpropinput_nchw.py [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+    log_device_placement=False,
+    inter_op_parallelism_threads=1)
+
+
+@pytest.mark.parametrize("padding", ("VALID", "SAME"))
+def test_conv2dbackpropinput_nchw(padding):
+    np_filter = np.random.rand(*filter_size_hwio).astype('f')
+    n_np_out = np.random.rand(*out_backprop_in_sizes[padding]).astype('f')
+    #Reshape to NHWC for TF
+    t_np_out = np.transpose(n_np_out, (0, 2, 3, 1))
```
Comment (kanvi-nervana) [inline test/python/bfloat16/test_conv2dbackpropinput_nchw.py]: You are transposing `out_backprop` to `NHWC` here as well https://github.com/tensorflow/ngraph-bridge/pull/183/files#diff-d7fb93f2ec39b43164446ba0e32fec3fR59
So this can be removed.

File: test/python/bfloat16/test_conv2d.py (hunk 6 above)
Comment (sindhu-nervana) [reply test/python/bfloat16/test_conv2d.py]: This is a special case which backend has so does not work for VALID or rather does not convert.

File: test/python/bfloat16/test_conv2d.py (hunk 7 above)
Comment (sindhu-nervana) [reply test/python/bfloat16/test_conv2d.py]: same comment as above

File: test/python/bfloat16/test_conv2dbackpropinput_nchw.py (hunk 8 above)
Comment (sindhu-nervana) [reply test/python/bfloat16/test_conv2dbackpropinput_nchw.py]: no, both are different. one is input shape and the other is np array shape with random data. so if i remove one of the transpose, feed_dict will fail with incorrect shapes.
